client/gateway_client.go

```diff
@@ -0,0 +1,208 @@
+package client
+
+import (
+	"context"
+	"fmt"
+
+	cryptotypes "github.com/cosmos/cosmos-sdk/crypto/types"
+	"github.com/pokt-network/poktroll/pkg/polylog"
+	apptypes "github.com/pokt-network/poktroll/x/application/types"
+	servicetypes "github.com/pokt-network/poktroll/x/service/types"
+	sessiontypes "github.com/pokt-network/poktroll/x/session/types"
+	"golang.org/x/exp/slices"
+
+	sdk "github.com/pokt-network/shannon-sdk"
+)
+
+// GatewayClient contains functionality needed to sign
+// and validate relays on the Shannon protocol.
+//
+// It is used to:
+//   - Fetch onchain data for the Shannon protocol integration
+//   - Sign relays using ring signatures (with the gateway's private key)
+//   - Validate relay responses
+//
+// It contains:
+//   - A full node for fetching onchain data (caching or just-in-time)
+//   - A relay signer for signing relays using ring signatures
+//   - The gateway address
+type GatewayClient struct {
+	shannonFullNode
+	*relaySigner
+	gatewayAddress string
+}
+
+func NewGatewayClient(
+	logger polylog.Logger,
+	fullNodeConfig FullNodeConfig,
+	gatewayAddress string,
+	gatewayPrivateKeyHex string,
+) (*GatewayClient, error) {
+	fullNode, err := getFullNode(logger, fullNodeConfig)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create Shannon full node: %v", err)
+	}
+
+	relaySigner := &relaySigner{
+		PrivateKeyHex:    gatewayPrivateKeyHex,
+		PublicKeyFetcher: fullNode,
+	}
+
+	return &GatewayClient{
+		shannonFullNode: fullNode,
+		relaySigner:     relaySigner,
+		gatewayAddress:  gatewayAddress,
+	}, nil
+}
+
+// FullNode is the interface that wraps the basic methods used to interface with the Shannon full node.
+// It is used to fetch onchain data for the Shannon protocol integration.
+//
+// It is implemented by the structs:
+//   - fullNode: default implementation of a full node for the Shannon.
+//   - fullNodeWithCache: the default full node with a SturdyC-based cache.
+type shannonFullNode interface {
```

Comment:
Not sure this should be private: it can be a public interface, which can be implemented by the user, or conveniently via other structs in the SDK.

Reply:
Ditto.